File: 0050-powx-n/0050-powx-n.java
class Solution {
    private double pow(double x, int n) {
        if(n == 0) return 1;
        double half = pow(x,n/2);
        return (n % 2 == 0) ? half * half : x * half * half;
    }

    public double myPow(double x, int n) {
        return (n < 0) ? 1/pow(x,-n) : pow(x,n);
    }
}

File: 0050-powx-n/0050-powx-n.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n == 0:
            return 1.0
        
        # Handle negative powers
        if n < 0:
            x = 1 / x
            n = -n
        
        result = 1.0
        current_product = x
        
        while n > 0:
            if n % 2 == 1:  # if current bit is 1
                result *= current_product
            current_product *= current_product  # square for next bit
            n //= 2
        
        return result
        